### DISCUSSION TITLE
Build and test automatically

### DISCUSSION THREAD

**Discussion by CodedOre** [CodedOre/NewCaw#45] [General]
I'm thinking about adding GitHub Actions to automatically build and test the code on this repository.

So far, I would want to add the following action:
- Building a Flatpak bundle
- (Maybe) Upload the build bundle to an flatpakrepo
- Test building the program with only one platform.
  Since you can (as of now) disable certain platforms, we will need to check that it still compiles when a platform is disabled, where automatic builds could support us quite a lot.

Thanks to the [Flatpak Builder Action](https://github.com/marketplace/actions/flatpak-builder), this should be doable.
There are a few things to discuss though:
- As of now, I do not include the OAuth key required for Twitter. In order for Flatpak Builder to build it, we would need to insert the key before running the build. Since we can run any bash command we want with an Github action, we may could do this with `sed`? Or we just include the key in the flatpak manifest.
- Should we want to provide an updatable development build, where to upload it to? Flathub is only for stable or beta software, both don't apply as of now.
- Am I overthinking the need for automated builds and we don't need them actually?

**Comment by IBBoard** [comment]
Tests on branches to make sure that they don't break things is a good practice. Cawbird has always been a bit… sparse on its tests! It's very rare I run them. But we should. I've got far more comprehensive tests on some of my other projects.

Building `main` into a bleeding edge package could be useful in some situations. It's simpler than my current "unstable" approach, which is to occasionally create patches and put them into the Open Build System, which I do as a manual process using the web interface. It would mean that we're stuck with Flatpak builds rather than proper native builds, though 😉

Does the package need to be uploaded anywhere? Can we just keep the latest one on GitHub and manually download and install? Or does Flatpak not work that way?

With the OAuth key, I've used GitHub Secrets to hold the value and then you can use `sed` and similar commands to edit it into a file (assuming we can't make it detect it from an environment variable).

**Reply by CodedOre** [reply]
> Tests on branches to make sure that they don't break things is a good practice. Cawbird has always been a bit… sparse on its tests! It's very rare I run them. But we should. I've got far more comprehensive tests on some of my other projects.

Yes, I also should look into extending the tests I currently have...

> Building `main` into a bleeding edge package could be useful in some situations. It's simpler than my current "unstable" approach, which is to occasionally create patches and put them into the Open Build System, which I do as a manual process using the web interface. It would mean that we're stuck with Flatpak builds rather than proper native builds, though wink

Not necessarily. GitHub Actions can run a lot, so we could also build and package those native ones. To be honest, I just focused on Flatpak because it's the "native" package format my distribution, Fedora Silverblue, uses.

> Does the package need to be uploaded anywhere? Can we just keep the latest one on GitHub and manually download and install? Or does Flatpak not work that way?

The Flatpak-Builder Action creates an Flatpak bundle that can be downloaded and installed. It's just that the package manager can't provide updates if there's no repository behind it. Should be the same issue with native packages as well. The question here is more if there's a case for updating dev-builds that would make this question relevant.

**Reply by IBBoard** [reply]
I guess there's nothing stopping us building other packages in GitHub actions. But it does feel like something that it's not quite set up for and that Open Build Service is better suited for 🙂 

I think "download and run a one-off build" has a decent use case for testing and confirming issues. The fact that there's no update repo is almost a benefit, because it should (hopefully) put people off using it unless they're testing a version.
